internal/wasm/jit/jit_amd64.go

```diff
@@ -4999,211 +3958,94 @@ func (c *amd64Compiler) returnFunction() error {
 		// "rb.caller" is BELOW the top address. See the above example for detail.
-		readReturnStackBasePointer.From.Offset = -(callFrameDataSize - callFrameReturnStackBasePointerOffset)
-		c.addInstruction(readReturnStackBasePointer)
-
-		putReturnStackBasePointer := c.newProg()
-		putReturnStackBasePointer.As = x86.AMOVQ
-		putReturnStackBasePointer.To.Type = obj.TYPE_MEM
-		putReturnStackBasePointer.To.Reg = reservedRegisterForCallEngine
-		putReturnStackBasePointer.To.Offset = callEngineValueStackContextStackBasePointerOffset
-		putReturnStackBasePointer.From.Type = obj.TYPE_REG
-		putReturnStackBasePointer.From.Reg = tmpRegister
-		c.addInstruction(putReturnStackBasePointer)
-	}
+		callFrameStackTopAddressRegister, -(callFrameDataSize - callFrameReturnStackBasePointerOffset),
+		tmpRegister,
+	)
+	c.compileRegisterToMemoryInstruction(x86.AMOVQ,
+		tmpRegister, reservedRegisterForCallEngine, callEngineValueStackContextStackBasePointerOffset)
 
 	// 2) Jump into the address of "ra.caller".
-	{
-		readReturnAddress := c.newProg()
-		readReturnAddress.As = x86.AMOVQ
-		readReturnAddress.To.Type = obj.TYPE_REG
-		readReturnAddress.To.Reg = tmpRegister
-		readReturnAddress.From.Type = obj.TYPE_MEM
-		readReturnAddress.From.Reg = callFrameStackTopAddressRegister
+	c.compileMemoryToRegisterInstruction(x86.AMOVQ,
 		// "ra.caller" is BELOW the top address. See the above example for detail.
-		readReturnAddress.From.Offset = -(callFrameDataSize - callFrameReturnAddressOffset)
-		c.addInstruction(readReturnAddress)
+		callFrameStackTopAddressRegister, -(callFrameDataSize - callFrameReturnAddressOffset),
+		tmpRegister,
+	)
 
-		jmpToReturnAddress := c.newProg()
-		jmpToReturnAddress.As = obj.AJMP
-		jmpToReturnAddress.To.Type = obj.TYPE_REG
-		jmpToReturnAddress.To.Reg = tmpRegister
-		c.addInstruction(jmpToReturnAddress)
-	}
+	c.compileJumpToRegister(tmpRegister)
 
 	// They were temporarily used, so we mark them unused.
 	c.locationStack.markRegisterUnused(regs...)
 	return nil
 }
 
-// callGoFunction adds instructions to call a Go function whose address equals the addr parameter.
-// jitStatus is set before making call, and it should be either jitCallStatusCodeCallBuiltInFunction or
-// jitCallStatusCodeCallHostFunction.
-func (c *amd64Compiler) callGoFunction(jitStatus jitCallStatusCode, index wasm.Index) error {
-	if jitStatus == jitCallStatusCodeCallBuiltInFunction {
-		// Set the functionAddress to the callEngine.exitContext functionCallAddress.
-		setFunctionCallAddress := c.newProg()
-		setFunctionCallAddress.As = x86.AMOVL
-		setFunctionCallAddress.From.Type = obj.TYPE_CONST
-		setFunctionCallAddress.From.Offset = int64(index)
-		setFunctionCallAddress.To.Type = obj.TYPE_MEM
-		setFunctionCallAddress.To.Reg = reservedRegisterForCallEngine
-		setFunctionCallAddress.To.Offset = callEngineExitContextBuiltinFunctionCallAddressOffset
-		c.addInstruction(setFunctionCallAddress)
-	}
+// TODO: this is deletable.
```

Comment:
nit: the comment on why this is deletable might help. Maybe it means `compileCallHostFunction` can be just replaced with `compileCallHostFunction`?

Reply:
I intended to solve this shortly, so I choose to leave as-is!